config/config.go

```diff
@@ -195,6 +204,74 @@ type Adapter struct {
 	Disabled bool `mapstructure:"disabled"`
 }
 
+// validateAdapterEndpoint validates to make sure that an adapter has a valid endpoint
+// associated with it
+func validateAdapterEndpoint(endpoint string, adapterName string, errs configErrors) configErrors {
+	if endpoint == "" {
+		errs = append(errs, fmt.Errorf("There's no default endpoint available for %s. Calls to this bidder/exchange will fail. "+
+			"Please set adapters.%s.endpoint in your app config", adapterName, adapterName))
+	} else {
+		// Check if endpoint has macros. If it does then resolve with dummy values
+		// for validation
+		if macros.ContainsMacro(endpoint) {
```

Comment:
Well, yes. BuildTemplate and ResolveMacros would still work and I agree that it won't affect performance because this is at start up. But, performing these two steps for each and every adapter endpoint/userSync URL which may or may not have macros to start with seems like not needed to me, no?

Reply:
Yeah, it's debatable. It'd cost the machine some cycles, but also reduce the [cyclomatic complexity](https://docs.codeclimate.com/docs/cyclomatic-complexity) of your function, which correlates to how many test cases you need to write to get full coverage.

My other worry is that the adapter code ultimately runs `macros.ResolveMacros` _without_ this check. So if Go's templates have some other restriction on string formats, you might actually skip some important validation here.

For example, suppose Go templates didn't allow `/` unless it were escaped like `\/`. If an adapter expects a template, but the value in the config doesn't have `{{ }}` vars, it might crash with an obscure error message because the validation here would miss it.